Cache intelligence v2 (#618)
harness/developer-hub: docs/continuous-integration/use-ci/caching-ci-data/cache-intelligence.md

---
title: Cache Intelligence

> description: You can set up your pipeline to cache your build artifacts automatically.

Reviewer: build dependencies

> description: You can set up your pipeline to cache your build artifacts automatically.

Reviewer 2: ```suggestion
description: Set up your pipelines to automatically cache build artifacts.
```

sidebar_position: 20
---

### Automatic caching

> You can set up your pipeline to cache your build artifacts automatically. Note the following:  

Reviewer 2: ```suggestion
You can set up your pipelines to automatically cache build artifacts. Note the following:  
```

> * Cache Intelligence is currently supported for the following. 

Reviewer 2: ```suggestion
* Cache Intelligence is currently supported for the following:
```

  - [Harness Cloud](/docs/continuous-integration/ci-quickstarts/hosted-builds-on-virtual-machines-quickstart) build infrastructure

>   - Basel, Maven, Gradle, Yarn, and Node frameworks

Reviewer 3: Bazel, not basel

Reviewer 3: Build tools is more accurate then frameworks here, but it should be ok

> * Harness Cloud can cache up to 2GB of data per account. All pipelines in the account can use the same cache. 

Reviewer 2: ```suggestion
* Harness Cloud can cache up to 2GB of data per account. All pipelines in the account use the same cache. 
```
"can use" kind of implies that it is configurable. If a pipeline uses cache intelligence, then it's using the account's cache.

* Cache retention window is 15 days, which resets whenever the cache gets updated.

> To enable Cache Intelligence on a CI Build stage, add the following lines to  the stage YAML definition: 

Reviewer 2: ```suggestion
To enable Cache Intelligence on a CI Build stage, add the following lines to the stage YAML definition: 
```

```yaml
    - stage:
        name: Build Jhttp
        identifier: Build_Jhttp
        description: ""
        type: CI
        spec:
          caching:            # --------------- ADD LINE
            enabled: true     # ----------------ADD LINE

>             path:             # optional paths to support specific frameworks

Reviewer: I think we should show one YAML sample with only caching.enabled=true. The paths can be mentioned in a separate YAML example indicating that the field can be used for frameworks where automatic caching is not supported.

Author: @hemanthmantri Removing the `path:` field from this example. The Save and Restore Cache docs for [S3 ](https://ci.pr.harness.io/pr-624/docs/continuous-integration/use-ci/caching-ci-data/saving-cache) and [GCS](https://ci.pr.harness.io/pr-624/docs/continuous-integration/use-ci/caching-ci-data/save-cache-in-gcs) include YAML samples. If these need updating, we can do this in a separate ticket.

              - /root/.mvn       # - absolute path if outside /harness 
              - vendor           # - relative path -- i.e., /harness/vendor
          cloneCodebase: true
```

> The Cache Intelligence API includes the following calls. 
> 

Reviewer 2: ```suggestion
###Cache Intelligence API

Use the Cache Intelligence API to get information about the cache or delete the cache.

```

Suggested breaking with a heading since the API calls aren't triggering a cache event (so not really related to the above heading, "Automated caching"). I also suggested rewriting the first sentence since there is a note between the sentence and the first call subsection.

:::note

You must have an API Key with [core_account_edit](/docs/platform/Role-Based-Access-Control/ref-access-management/api-permissions-reference#harness-api-permissions) permissions to invoke these APIs.

For information about API keys, go to [Add and Manage API Keys](/docs/platform/role-based-access-control/add-and-manage-api-keys). 

:::

#### Get cache metadata

Get metadata about the cache, such as the size and path. 

```
curl --location --request GET 'https://app.harness.io/gateway/ci/cache/info?accountIdentifier=$HARNESS_ACCOUNT_ID' \
--header 'Accept: application/json' \
--header 'Authorization: Bearer $AUTH_TOKEN'
```

> #### Delete cache for pipeline

Reviewer 2: The subsection heading is "delete cache for pipeline", but I think it's deleting the entire account's cache.

> Delete the cache. You can include an optional `path` parameter to delete a specific subfolder in the cache.

Reviewer: sub-directory?

```

> curl --location --request DELETE 'https://app.harness.io/gateway/ci/cache/info?accountIdentifier=$HARNESS_ACCOUNT_ID&path=/path/to/deleted/folder' \

Reviewer: /path/to/delete
@jamie-harness : Can we specify multiple directories to delete?

Reviewer 2: Editorial comment: I would probably format the last part as `&path=<path to directory to delete>` or something. I completely overlooked it in the current format assuming it was a standard file path.

---
title: Cache Intelligence
description: Set up your pipeline to cache build dependencies automatically.
sidebar_position: 20
---

### Automatic caching

You can set up your pipeline to cache build dependencies automatically. Note the following:  

* Cache Intelligence is currently supported for the following:
  - [Harness Cloud](/docs/continuous-integration/ci-quickstarts/hosted-builds-on-virtual-machines-quickstart) build infrastructure
  - Bazel, Maven, Gradle, Yarn, and Node build tools

>   - The build tool must store cache data in the default location for that tool.  

Reviewer 2: Should this bullet be indented more or outdented?

* Harness Cloud can cache up to 2GB of data per account. All pipelines in the account use the same cache. 
* Cache Intelligence stores cache data in the `/harness` folder by default. 
* Cache retention window is 15 days, which resets whenever the cache gets updated.

To enable Cache Intelligence on a CI Build stage, add the following lines to the stage YAML definition: 

```yaml
    - stage:
        name: Build Jhttp
        identifier: Build_Jhttp
        description: ""
        type: CI
        spec:
          caching:            # --------------- ADD LINE
            enabled: true     # ----------------ADD LINE
          cloneCodebase: true

> ```

Reviewer: We should have another YAML example for unsupported build tools and mention the `paths` key to be used to specify the list of paths that can be cached.

Author: Will implement this in a future PR: https://harness.atlassian.net/browse/DOC-2583
